Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp

```diff
@@ -0,0 +1,234 @@
+        enabledSitesMap().add(domain, EnhancedSecurityReason::InsecureLoad);
+}
+
+void EnhancedSecurityTracking::transfer(EnhancedSecurityTracking& other)
+{
+    m_activeState = other.m_activeState;
+    m_activeReason = other.m_activeReason;
+    m_initialProtectedDomain = other.m_initialProtectedDomain;
+}
+
+EnhancedSecurity EnhancedSecurityTracking::enhancedSecurityState() const
+{
+    if (m_activeState != ActivationState::Active)
+        return EnhancedSecurity::Disabled;
+
+    switch (enhancedSecurityReason()) {
+    case EnhancedSecurityReason::None:
+        ASSERT_NOT_REACHED();
+        return EnhancedSecurity::Disabled;
+
+    case EnhancedSecurityReason::InsecureProvisional:
+    case EnhancedSecurityReason::InsecureLoad:
+        return EnhancedSecurity::EnabledInsecure;
+
+    case EnhancedSecurityReason::Policy:
+        return EnhancedSecurity::EnabledPolicy;
+    }
+
+    ASSERT_NOT_REACHED();
+    return EnhancedSecurity::Disabled;
+}
+
+void EnhancedSecurityTracking::reset()
+{
+    m_activeState = ActivationState::None;
+    m_activeReason = EnhancedSecurityReason::None;
+}
+
+void EnhancedSecurityTracking::makeDormant()
+{
+    m_activeState = ActivationState::Dormant;
+}
+
+void EnhancedSecurityTracking::makeActive()
+{
+    m_activeState = ActivationState::Active;
+}
+
+static EnhancedSecurityReason reasonForEnhancedSecurity(EnhancedSecurity state)
+{
+    switch (state) {
+    case EnhancedSecurity::Disabled:
+        return EnhancedSecurityReason::None;
+
+    case EnhancedSecurity::EnabledInsecure:
+        return EnhancedSecurityReason::InsecureLoad;
+
+    case EnhancedSecurity::EnabledPolicy:
+        return EnhancedSecurityReason::Policy;
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+void EnhancedSecurityTracking::enableFor(EnhancedSecurityReason reason, API::Navigation& navigation)
+{
+    m_activeState = ActivationState::Active;
+    m_activeReason = reason;
+    m_initialProtectedDomain = RegistrableDomain(navigation.currentRequest().url());
+
+    enabledSitesMap().add(m_initialProtectedDomain, m_activeReason);
+}
+
+void EnhancedSecurityTracking::trackChangingSiteNavigation(API::Navigation& navigation)
+{
+    if (enhancedSecurityReason() == EnhancedSecurityReason::InsecureProvisional) {
+        m_activeReason = EnhancedSecurityReason::InsecureLoad;
+
+        if (enabledSitesMap().contains(m_initialProtectedDomain)) {
+            if (enabledSitesMap().get(m_initialProtectedDomain) == EnhancedSecurityReason::InsecureProvisional)
+                enabledSitesMap().set(m_initialProtectedDomain, EnhancedSecurityReason::InsecureLoad);
+        }
+    }
+}
+
+void EnhancedSecurityTracking::trackSameSiteNavigation(API::Navigation& navigation)
+{
+    bool isHttps = navigation.currentRequest().url().protocolIs("https"_s);
+
+    if (enhancedSecurityReason() == EnhancedSecurityReason::InsecureProvisional && isHttps) {
+        reset();
+
+        auto domain = RegistrableDomain { navigation.currentRequest().url() };
```

Comment:
` RegistrableDomain domain { navigation.currentRequest().url() };` is more common and more concise.